File: 2022/11/solution.pas
{ solution.pas -- monkey business -- Troy Brumley BlameTroi@gmail.com }

program solution;

{ expects to be compiled with the following two mode flags and command
line/fpc.cfg options of -Ciort -Sgix -Sh- for full checks, allowing
exceptions, goto, inlining, and prefer classic strings. }
{$LONGSTRINGS OFF}            { to be sure, use AnsiString or pchar explicitly }
{$BOOLEVAL ON}                { standard pascal expects this }


{ advent of code 2022 day 11 -- monkey business

  which monkeys have the most important stuff? }

type
   integer = int64;

const
   { input markers in expected order }
   cmonkey    = 'Monkey ';                        { n: }
   cstart     = '  Starting items: ';             { n (, n)* }
   coper      = '  Operation: new = old ';        { [+*] n }
   ctest      = '  Test: divisible by ';          { n }
   cift       = '    If true: throw to monkey ';  { n }
   ciff       = '    If false: throw to monkey '; { n }
   cold       = ' old';
{$define xxxTESTING }
{$ifdef TESTING }
   MAXITEMS   = 10;             { 10 test, 36 live }
   LASTMONKEY = 3;              { 0-3 test, 0-7 live }
{$else}
   MAXITEMS   = 36;             { 36 live }
   LASTMONKEY = 7;              { 0-7 live }
{$endif}
   MINROUNDS  = 20;             { for both test and live }
   MAXROUNDS  = 10000;          { for both test and live }
   CRELIEF    = 3;              { a scaling value for min rounds }


type
   tmonkey = record
                id          : char;        { will use index }
                numitems    : integer;      { how many items }
                priority    : array[1..MAXITEMS] of integer;
                operation   : char;        { + or * }
                opervalue   : integer;      { 0 means old, otherwise integer }
                testdivisor : integer;      { priority test value}
                throwtrue   : integer;      { index of true dest }
                throwfalse  : integer;      { index of false dest}
                examined    : integer;      { how many items examined }
             end;

var
   monkey : array [0..LASTMONKEY] of tmonkey;


{$I aocinput.p }
{$I str2int.p }
{$I strbegins.p }
{$I int2str.p }


{ initialize }

procedure init;

var
   i, j : integer;

begin
   for i := 0 to LASTMONKEY do
      with monkey[i] do begin
         id := ' ';
         numitems := 0;
         for j := 1 to MAXITEMS do
            priority[j] := 0;
         operation := ' ';
         opervalue := 0;
         testdivisor := 0;
         throwtrue := 0;
         throwfalse := 0;
         examined := 0;
      end;
end;


{ load and parse }

procedure load;

var
   i, j, p : integer;
   s       : string;

begin
   aocrewind;
   i := 0;
   while not aoceof do begin
      s := aocreadexpecting(cmonkey);
      with monkey[i] do begin
         id := s[length(cmonkey) + 1];

         s := aocreadexpecting(cstart);
         p := length(cstart) + 1;
         j := 0;
         while p < length(s) do begin
            j := j + 1;
            priority[j] := str2int(s, p);
            p := p + 2; { skip ', ' }
         end;
         numitems := j;

         s := aocreadexpecting(coper);
         operation := s[length(coper) + 1];
         p := length(coper) + 3; { skip past ' + ' }
         if s[p] in ['0'..'9'] then
            opervalue := str2int(s, p)
         else
            opervalue := 0; { signifies old priority }

         s := aocreadexpecting(ctest);
         p := length(ctest) + 1; { skip to digits }
         testdivisor := str2int(s, p);

         s := aocreadexpecting(cift);
         p := length(cift) + 1;
         throwtrue := str2int(s, p);

         s := aocreadexpecting(ciff);
         p := length(ciff) + 1;
         throwfalse := str2int(s, p);

         if not aoceof then
            s := aocreadexpecting('');
      end;
      i := i + 1;
   end;
end;


{ report on the monkeys and the items }

procedure report;

var
   i, j : integer;
   s    : string;

begin
   for i := 0 to LASTMONKEY do
      with monkey[i] do begin
         aoclog('');
         aoclog('monkey ' + id);
         s := 'holding ' + int2str(numitems) + ' of priorities [';
         for j := 1 to numitems do
            s := s + ' ' + int2str(priority[j]);
         s := s + ' ]';
         aoclog(s);
         s := 'operation is new = old ' + operation + ' ';
         if opervalue = 0 then
            s := s + 'old'
         else
            s := s + int2str(opervalue);
         aoclog(s);
         aoclog('test is divisible by ' + int2str(testdivisor));
         aoclog('when true throws to ' + int2str(throwtrue));
         aoclog('when false throws to ' + int2str(throwfalse));
         aoclog('items examined ' + int2str(examined));
      end;
   aoclog('');
   aoclogflush;
end;


{ the monkey wonders if you're worried? are you worried? the
  operation is applied and then the new priority is divided
  by the relief value and rounded down. }

procedure dooperation(m      : integer;  { monkey }
                      i      : integer;  { item number }
                      relief : integer); { how you feeling, bunky? }
var
   op, np : integer;

begin
   with monkey[m] do begin
      op := priority[i];
      if opervalue = 0 then
         np := op
      else
         np := opervalue;
      if operation = '+' then
         priority[i] := op + np
      else if operation = '*' then
         priority[i] := op * np
      else
         aochalt('*** error illegal operation ' + operation + ' for monkey ' + int2str(m));
      { take the mod of keep the value under the limit when doing prime multiples }
      if relief = CRELIEF then
         priority[i] := priority[i] div relief
      else
         priority[i] := priority[i] mod relief;
   end;
end;


{ mine! mine! add the incoming priority to the end of this
  monkey's list. }

procedure catchitem(m : integer;
                    n : integer);  { item priority value }
begin
   with monkey[m] do begin
      numitems := numitems + 1;
      priority[numitems] := n;
   end;
end;


{ the monkey makes up his mind. }

function dotest(m : integer;
                i : integer) : boolean;
begin
   with monkey[m] do
      dotest := (priority[i] mod testdivisor) = 0;
end;


{ calculate a relief value. i'm not much on number theory, but
  everything was a prime number so after thinking about it for
  a while i still had nothing and checked reddit. this is similar
  to the ghost walk problem in that we're dealing with prime
  multiples. }

function calculatedrelief : integer;

var
   i : integer;

begin
   calculatedrelief := 1;
   for i := 0 to LASTMONKEY do
      calculatedrelief := calculatedrelief * monkey[i].testdivisor;
end;

{ round and round we go. }

procedure goaround(rounds : integer;
                   relief : integer);

var
   i, j, k : integer;

begin
   for i := 1 to rounds do begin
      for j := 0 to LASTMONKEY do
         with monkey[j] do begin
            examined := examined + numitems;
            for k := 1 to numitems do
               dooperation(j, k, relief);
            for k := 1 to numitems do
               if dotest(j, k) then
                  catchitem(throwtrue, priority[k])
               else
                  catchitem(throwfalse, priority[k]);
            numitems := 0;
         end;
      aoclog('');
      aoclog('round ' + int2str(i));
      report;
   end;
end;


{ find the number of items examined by the monkeys who touched your
  stuff most frequently. bad monkeys. report the values of the top
  two multiplied together. }

function multiplytoptwo : integer;

var
   i, j, k : integer;

begin
   j := 0;  { index of biggest found }
   k := 1;  { index of second biggest found }
      if monkey[j].examined < monkey[k].examined then begin  { keep correctly ordered }
         i := j; j := k; k := i;
      end;
   for i := 2 to LASTMONKEY do begin
      if monkey[i].examined <= monkey[k].examined then       { too small }
         continue;
      if monkey[i].examined <= monkey[j].examined then       { between j and k }
         k := i
      else begin
         k := j;
         j := i;
      end;
   end;
   aoclog('top two at ' + int2str(j) + ' ' + int2str(k));
   multiplytoptwo := monkey[j].examined * monkey[k].examined;
end;


begin
   aocopen;
   aocnolog;
   init;
   load;
   report;
   goaround(MINROUNDS, CRELIEF);
   writeln('part one answer : ', multiplytoptwo);
   init;
   load;
   goaround(MAXROUNDS, calculatedrelief);
   writeln('part two answer : ', multiplytoptwo);
   aocclose;
end.

{ end solution.pas }

File: 2022/11/solution.f90
! solution.f90 -- 2022 day 11 -- monkey business -- troy brumley blametroi@gmail.com

! advent of code 2022 day 11 -- monkey business -- which monkeys have your stuff?
!
! all your stuff fell out of the your pack and monkeys are tossing it around. the
! monkeys pay attention to your anxiety level, devious monkeys.

program solution

   use iso_fortran_env, only: int64
   use aocinput, only: open_aoc_input, read_aoc_input, close_aoc_input, max_aoc_reclen, rewind_aoc_input

   implicit none

   integer, parameter            :: AOCIN = 10
   integer(kind=int64)           :: part_one, part_two  ! results
   character(len=max_aoc_reclen) :: rec

   ! input should be a block for each monkey as follows.
   character(len=*), parameter :: cmonkey = "Monkey "                      ! n: (0-9)
   integer, parameter :: lcmonkey = 7
   character(len=*), parameter :: cstart = "  Starting items: "            ! n (, n)*
   integer, parameter :: lcstart = 18
   character(len=*), parameter :: coper = "  Operation: new = old "        ! [+*] n
   integer, parameter :: lcoper = 23
   character(len=*), parameter :: ctest = "  Test: divisible by "          ! n
   integer, parameter :: lctest = 21
   character(len=*), parameter :: cift = "    If true: throw to monkey "   ! n
   integer, parameter :: lcift = 29
   character(len=*), parameter :: ciff = "    If false: throw to monkey "  ! n
   integer, parameter :: lciff = 30
   character(len=*), parameter :: cold = " old"
   integer, parameter :: lcold = 4

   ! constraints or limits
   integer, parameter             :: MAXITEMS = 36         ! 10 test, 36 live
   integer, parameter             :: LASTMONKEY = 7        ! 0-3 test, 0-7 live
   integer, parameter             :: MINROUNDS = 20        ! for both test and live
   integer, parameter             :: MAXROUNDS = 10000     ! for both test and live
   integer(kind=int64), parameter :: CRELIEF = 3           ! a scaling value for min rounds

   ! runtime representation of the monkey and items
   type monkey_t
      character(len=1)              :: id              ! we'll actually use the subscript to id these
      integer                       :: numitems        ! how many
      integer(kind=int64), dimension(1:MAXITEMS) :: priority
      character(len=1)              :: operation       ! multiply or add
      integer                       :: opervalue       ! 0 means old, otherwise integer
      integer                       :: testdivisor     ! priority test value
      integer                       :: throwtrue       ! send there if true
      integer                       :: throwfalse      ! otherwise here
      integer(kind=int64)           :: examined        ! how many items has the monkey examined
   end type monkey_t

   type(monkey_t), dimension(0:LASTMONKEY) :: monkey

   rec = ""
   call open_aoc_input(AOCIN)

   ! initialize
   part_one = 0; part_two = 0

   call init
   call load
   call goaround(MINROUNDS, CRELIEF)
   part_one = multiplytoptwo()

   call init
   call load
   call goaround(MAXROUNDS, calculatedrelief())
   part_two = multiplytoptwo()

   ! report and close
   print *
   print *, part_one, "part one"
   print *, part_two, "part two"
   print *

   call close_aoc_input(AOCIN)

contains

   subroutine init
      implicit none
      integer :: i
      do i = 0, LASTMONKEY
         monkey(i)%id = " "
         monkey(i)%numitems = 0
         monkey(i)%priority = 0
         monkey(i)%operation = ' '
         monkey(i)%opervalue = 0
         monkey(i)%testdivisor = 0
         monkey(i)%throwtrue = 0
         monkey(i)%throwfalse = 0
         monkey(i)%examined = 0
      end do
   end subroutine init

   ! read a record, but it must match the epexted marker or we'll error out
   subroutine read_expecting(marker, eofok)
      implicit none
      character(len=*), intent(in) :: marker
      logical, intent(in)          :: eofok
      logical                      :: readok
      readok = read_aoc_input(AOCIN, rec)
      if ((readok) .and. (rec(1:len(marker)) == marker)) return
      rec = "$$$EOF$$$"
      if (eofok) return
      print *, "*** error *** unexpected eof on input"
      stop 1
   end subroutine

   subroutine load
      implicit none
      integer :: i, j
      integer, allocatable :: p(:)
      call rewind_aoc_input(AOCIN)
      i = 0
      do
         call read_expecting(cmonkey, .true.)
         if (trim(rec) == "$$$EOF$$$") exit

         monkey(i)%id = rec(lcmonkey + 1:lcmonkey + 1)
         call read_expecting(cstart, .false.)
         p = integers_from_string(rec, lcstart + 1, ",")
         do j = 1, size(p)
            monkey(i)%priority(j) = p(j)
         end do
         monkey(i)%numitems = size(p)
         call read_expecting(coper, .false.)
         monkey(i)%operation = rec(lcoper + 1:lcoper + 1)
         monkey(i)%opervalue = 0 ! 0 means 'old value'
         if (rec(lcoper + 3:lcoper + 6) /= "old") then
            read (rec(lcoper + 3:len(rec)), *) monkey(i)%opervalue
         end if
         call read_expecting(ctest, .false.)
         read (rec(lctest + 1:len(rec)), *) monkey(i)%testdivisor
         call read_expecting(cift, .false.)
         read (rec(lcift + 1:len(rec)), *) monkey(i)%throwtrue
         call read_expecting(ciff, .false.)
         read (rec(lciff + 1:len(rec)), *) monkey(i)%throwfalse
         call read_expecting("", .true.)

         i = i + 1
         ! print *, i, monkey(i)%numitems, monkey(i)%operation, monkey(i)%opervalue, monkey(i)%testdivisor, &
         !    monkey(i)%throwtrue, monkey(i)%throwfalse
      end do
   end subroutine load

   function integers_from_string(str, beg, sep) result(a)
      ! a beautiful find from a SO question. author is Vladimir Fuka, LadaF at github.
      ! needs error checking for general use, and i got strict on intent so the input
      ! string needs a local copy, but it's still a cool idea.
      implicit none
      integer, allocatable       :: a(:)
      character(*), intent(in)   :: str    ! string to pluck integers from
      integer, intent(in)        :: beg    ! start pos in string, 1 based
      character(*), intent(in)   :: sep    ! separator character, only first pos used
      character(len(str))        :: cpy
      integer                    :: i, n
      n = 0
      cpy = str(beg:len(str))
      do i = 1, len_trim(cpy)
         if (cpy(i:i) == sep) then
            n = n + 1
            cpy(i:i) = "," ! read wants a comma, your sep can be anything
         end if
      end do
      allocate (a(n + 1))
      read (cpy, *) a
   end function integers_from_string

   ! the monkey wonders if you're worried? are you worried? the operation is
   ! applied and then the new priority is divided by the relief value and rounded
   ! down.

   subroutine dooperation(m, i, relief)
      implicit none
      integer, intent(in) :: m       ! the monkey
      integer, intent(in) :: i       ! item number
      integer(kind=int64), intent(in) :: relief  ! how you feeling?

      integer(kind=int64) :: op, np

      op = monkey(m)%priority(i)
      if (monkey(m)%opervalue == 0) then
         np = op
      else
         np = monkey(m)%opervalue
      end if
      if (monkey(m)%operation == "+") then
         monkey(m)%priority(i) = op + np
      else if (monkey(m)%operation == "*") then
         monkey(m)%priority(i) = op*np
      else
         print *, "*** error *** illegal operation '", monkey(m)%operation, "' for monkey", m
         stop 1
      end if
      ! take the mod instead of doing division when doing the prime multiples version
      if (relief == CRELIEF) then
         monkey(m)%priority(i) = monkey(m)%priority(i)/relief
      else
         monkey(m)%priority(i) = mod(monkey(m)%priority(i), relief)
      end if
   end subroutine

   ! mine! mine! add the incoming priority to the end of this monkey's list.
   subroutine catchitem(m, n)
      implicit none
      integer, intent(in)             :: m   ! which monkey
      integer(kind=int64), intent(in) :: n   ! item priority value
      monkey(m)%numitems = monkey(m)%numitems + 1
      monkey(m)%priority(monkey(m)%numitems) = n
   end subroutine

   ! the monkey makes up his mind.
   function dotest(m, i)
      implicit none
      integer, intent(in) :: m   ! which monkey
      integer, intent(in) :: i   ! which item
      logical             :: dotest
      dotest = mod(monkey(m)%priority(i), monkey(m)%testdivisor) == 0
   end function

   ! calculate a relief value. i'm not much on number theory, but everything was
   ! a prime number so after thinking about it for a while i still had nothing
   ! and checked reddit. this is similar to the ghost walk problem in that we're
   ! dealing with prime multiples.
   function calculatedrelief() result(res)
      implicit none
      integer(kind=int64) :: res
      integer             :: i
      res = 1
      do i = 0, LASTMONKEY
         res = res*monkey(i)%testdivisor
      end do
   end function calculatedrelief

   ! round and round we go, and you're the monkey in the middle
   subroutine goaround(rounds, relief)
      implicit none
      integer, intent(in) :: rounds
      integer(kind=int64), intent(in) :: relief
      integer :: i, j, k
      do i = 1, rounds
         do j = 0, LASTMONKEY
            monkey(j)%examined = monkey(j)%examined + monkey(j)%numitems
            do k = 1, monkey(j)%numitems
               call dooperation(j, k, relief)
            end do
            do k = 1, monkey(j)%numitems
               if (dotest(j, k)) then
                  call catchitem(monkey(j)%throwtrue, monkey(j)%priority(k))
               else
                  call catchitem(monkey(j)%throwfalse, monkey(j)%priority(k))
               end if
            end do
            monkey(j)%numitems = 0
         end do
      end do
   end subroutine goaround

   ! find the number of items examined by the monkeys who touched your stuff
   ! most frequently. bad monkeys. report the values of the top two multiplied
   ! together. the things we do to avoid sorting or creating a priority queue.
   function multiplytoptwo() result(res)
      implicit none
      integer(kind=int64) :: res
      integer :: i, j, k

      j = 0 ! subscript of biggest found
      k = 1 ! subscript of second biggest found
      ! we've arbitrarily picked the first two, now make sure they're in order
      if (monkey(j)%examined < monkey(k)%examined) then
         i = j; j = k; k = i    ! ordered correctly
      end if
      ! now find the real top two
      do i = 2, LASTMONKEY
         if (monkey(i)%examined <= monkey(k)%examined) cycle ! too small
         if (monkey(i)%examined <= monkey(j)%examined) then ! between j and k
            k = i
         else                                               ! new top
            k = j
            j = i
         end if
      end do
      res = monkey(j)%examined*monkey(k)%examined
   end function multiplytoptwo

end program solution
